packages/@react-aria/button/src/useButton.ts

```diff
@@ -101,6 +106,31 @@ export function useButton(props: AriaButtonOptions<ElementType>, ref: RefObject<
     focusableProps.tabIndex = isDisabled ? -1 : focusableProps.tabIndex;
   }
   let buttonProps = mergeProps(focusableProps, pressProps, filterDOMProps(props, {labelable: true}));
+  
+  useEffect(() => {
+    function onPointerDown() {
+      setClickByUser(true);
+      setPreventFocusOnPress(false);
+    }
+    function onClick(e: MouseEvent) {
+      let button = (e.currentTarget as HTMLButtonElement);
+      // this if-statement is to test whether the click event was generated by an implicit form submission triggered by the user pressing `Enter` key
+      // more info on this: https://html.spec.whatwg.org/#implicit-submission
+      if (button.type === 'submit' && button.form && !clickByUser) {
+        setPreventFocusOnPress(true);
+      }
+      setClickByUser(false);
+    }
+    if (ref?.current && ref.current instanceof HTMLButtonElement) {
+      let button =  ref.current;
+      button.addEventListener('pointerdown', onPointerDown);
+      button.addEventListener('click', onClick);
+      return () => {
+        button.removeEventListener('pointerdown', onPointerDown);
```

Comment:
instead of needing a useEffect to hook up the events, I think we should be able to just pass them into the mergeProps down below, there's already a 'click' listener there

Reply:
Ha this is weird.. 

But first of all, thanks for the comment! I also think passing these callbacks in `mergeProps()` in the return statement is a better idea, and so I went ahead and tried to do just that, only to find myself puzzled by this weird behavior.. that I'm about to describe:

The implicit form submission (a.k.a, pressing the Enter key) is not detected and corrected in first try; by first try I mean when the input first gains (or re-gains) the focus. On the other hand, any subsequent tries (after the first failure) seem to work fine. Here's a video:

https://github.com/adobe/react-spectrum/assets/71210554/5a4e583d-a2eb-40dc-926d-656925bc00e4

In code, it looks something like this:
```ts
// In useButton()

...

let [clickByUser, setClickByUser] = useState(false);
let [preventFocusOnPress, setPreventFocusOnPress] = useState(false);

let detectClickByUser = useCallback((e: PointerEvent) => {
  if (e.currentTarget instanceof HTMLButtonElement) {
    setClickByUser(true);
    setPreventFocusOnPress(false);
  }
}, []);

let detectImplicitFormSubmission = useCallback((e: MouseEvent) => {
  if (e.currentTarget instanceof HTMLButtonElement) {
    let button = e.currentTarget;
    // this if-statement is to test whether the click event was generated by an implicit form submission triggered by the user pressing `Enter` key
    // more info on this: https://html.spec.whatwg.org/#implicit-submission
    if (button.type === 'submit' && button.form && !clickByUser) {
      setPreventFocusOnPress(true);
    }
    setClickByUser(false);
  }
}, [clickByUser]);

let {pressProps, isPressed} = usePress({
  onPressStart,
  onPressEnd,
  onPressChange,
  onPress,
  onPressUp,
  isDisabled,
  preventFocusOnPress: preventFocusOnPressProps !== undefined ? preventFocusOnPressProps : preventFocusOnPress,
  ref
});

...

return {
  isPressed, // Used to indicate press state for visual
  buttonProps: mergeProps(
    {
      onPointerDown: detectClickByUser,
      onClick: detectImplicitFormSubmission
    },
    additionalProps, 
    buttonProps, {
      'aria-haspopup': props['aria-haspopup'],
      'aria-expanded': props['aria-expanded'],
      'aria-controls': props['aria-controls'],
      'aria-pressed': props['aria-pressed'],
      onClick: (e) => {
        if (deprecatedOnClick) {
          deprecatedOnClick(e);
          console.warn('onClick is deprecated, please use onPress');
        }
      }
    }
  )
};
```

I've also tried placing state updaters (`setClickByUser()` and `setPreventFocusOnPress()` inside `flushSync()`, but to no avail. 

Also placing the callbacks as the first argument to `mergeProps()` (since `chain()` calls the callbacks in the order of their appearance) doesn't seem to have any effect as well.

I'm honestly clueless as to why this is happening with the above change.. so any help here to shed light on the root cause of this issue would be much appreciated 🙏.

Reply:
This is mostly a wild speculation, but I suspect the above code doesn't work because we're trying to _update_ (in `detectImplicitFormSubmission()`) and _access_ (in `pressProps.onClick()`) the state value (`preventFocusOnPress`) in a single rendering path. 

On the other hand, if we register our own handlers (bypassing React's event-managing system) in `useEffect()` then when an event occurs somehow the click handler in the `useEffect()` is called first, which results in a state update, React detects the update and updates the handler to have most current state value before it fires an event to the handler it manages (i.e., `pressProps.onClick()`).

Not sure if there's a way around not resorting to using `useEffect()` in this case.